packages/tpetra/core/src/Tpetra_Details_Behavior.cpp

```diff
@@ -22,6 +24,89 @@ namespace { // (anonymous)
     return s;
   }
 
+  template<class Func>
+  void
+  split(const std::string& s, Func f, const char sep=':')
+  {
+    typedef std::string::size_type size_type;
+    size_type cur_pos, last_pos=0, length=s.length();
+    while(last_pos < length + 1)
+    {
+      cur_pos = s.find_first_of(sep, last_pos);
+      if(cur_pos == std::string::npos)
+      {
+        cur_pos = length;
+      }
+      if(cur_pos!=last_pos) {
+        auto token = std::string(s.data()+last_pos, (size_type)cur_pos-last_pos);
+        f(token);
+      }
+      last_pos = cur_pos + 1;
+     }
+    return;
+  }
+
+  /// \brief Determine if environment variable is a boolean or not
+  ///
+  /// If the environment variable is "true", return 1
+  /// If the environment variable is "false", return 0
+  /// Otherwise, return -1
+  int
+  environmentVariableState(std::string environmentVariableValue)
+  {
+    int state = -1;
+    std::string v = stringToUpper(environmentVariableValue);
+    if      (v == "1" || v == "YES" || v == "TRUE"  || v == "ON")
+      // Environment variable is "ON"
+      return 1;
+    else if (v == "0" || v == "NO"  || v == "FALSE" || v == "OFF")
+      // Environment variable is "OFF"
+      return 0;
+    // Environment has some other non-boolean value
+    return -1;
+  }
+
+  void
+  getEnvironmentVariableAsMap (const char environmentVariableName[],
+                               std::map<std::string, bool>& valsMap,
+                               const bool defaultValue)
+  {
+    using std::string;
+
+    const char* varVal = std::getenv (environmentVariableName);
+    if (varVal == NULL) {
+      // Apply default value to all
+      valsMap.clear();
+      valsMap["ALL"] = defaultValue;
+      return;
+    }
+
+    // Variable is not empty.
+    const string varStr(varVal);
+    std::vector<string> names;
+    split(varStr, [&](string x){names.push_back(x);});
+    for (auto name: names) {
+      int state = environmentVariableState(name);
+      if (state == 1) {
+        // Global value takes precedence
```

Comment:
I would say, the global value is the default value for all classes.  However, users may still turn on or off the value for particular classes.  This would imply returning the global default value separately, not as part of the `std::map`.

Reply:
The behavior is almost like you suggest already.  The only case where there is ambiguity is if, say `TPETRA_VERBOSE` is specified as:

```TPETRA_VERBOSE=Distributor:CrsMatrix:ON```

The `ON` value would trigger `VERBOSE` to be `ON` for all classes.

Reply:
Ah, OK then.  That's fine.  Silly of the user, but fine ;-) .